File: samples/sample.ml
(* Define a module with a few functions *)
module MyModule = struct
  let add x y = x + y

  let sub x y = x - y

  let mul x y = x * y

  let div x y = 
    if y = 0 then 
      None 
    else 
      Some (x / y)
end

(* Record type *)
type person = {
  name : string;
  age : int;
}

(* Variant type *)
type color =
  | Red
  | Green
  | Blue

(* Recursive function *)
let rec factorial n =
  if n <= 1 then 1 else n * factorial (n - 1)

(* Pattern matching example *)
let describe_color c =
  match c with
  | Red -> "Red"
  | Green -> "Green"
  | Blue -> "Blue"

(* Option type example *)
let divide a b =
  if b = 0 then None else Some (a / b)

(* List processing with higher-order functions *)
let numbers = [1; 2; 3; 4; 5]

let squares = List.map (fun x -> x * x) numbers

let evens = List.filter (fun x -> x mod 2 = 0) numbers

(* Using the Printf module for formatted output *)
let () =
  Printf.printf "Factorial of 5: %d\n" (factorial 5);
  Printf.printf "Squares of numbers: %s\n"
    (String.concat "; " (List.map string_of_int squares));
  Printf.printf "Even numbers: %s\n"
    (String.concat "; " (List.map string_of_int evens))

(* String literals and variable bindings *)
let foobar = "Lorem Ipsum"
let foo = "123"
let bar = 42
let baz = 3.14

let () =
  Printf.printf "foobar: %s\n" foobar;
  Printf.printf "foo: %s\n" foo;
  Printf.printf "bar: %d\n" bar;
  Printf.printf "baz: %.2f\n" baz

(* Control structures *)
let () =
  if bar > 10 then
    Printf.printf "bar is greater than 10\n"
  else
    Printf.printf "bar is not greater than 10\n"

(* Loops *)
let () =
  for i = 0 to 4 do
    Printf.printf "for loop iteration: %d\n" i
  done;

  let i = ref 0 in
  while !i < 5 do
    Printf.printf "while loop iteration: %d\n" !i;
    incr i
  done

(* Using the MyModule functions *)
let () =
  let open MyModule in
  Printf.printf "10 + 5 = %d\n" (add 10 5);
  Printf.printf "10 - 5 = %d\n" (sub 10 5);
  Printf.printf "10 * 5 = %d\n" (mul 10 5);
  match div 10 2 with
  | Some result -> Printf.printf "10 / 2 = %d\n" result
  | None -> Printf.printf "Division by zero\n"

File: samples/sample.zig
const std = @import("std");

pub fn main() void {
    const stdout = std.io.getStdOut().writer();
    var foobar = 42;
    var foo = "Lorem Ipsum";
    var bar = 123;
    var baz: f64 = 3.14;
    var flag = true;

    stdout.print("foobar: {}\n", .{foobar}) catch {};
    stdout.print("foo: {}\n", .{foo}) catch {};
    stdout.print("bar: {}\n", .{bar}) catch {};
    stdout.print("baz: {}\n", .{baz}) catch {};
    stdout.print("flag: {}\n", .{flag}) catch {};

    // Control Structures
    if (foobar > bar) {
        stdout.print("foobar is greater than bar\n", .{}) catch {};
    } else {
        stdout.print("foobar is not greater than bar\n", .{}) catch {};
    }

    // Loops
    for (bar) |i| {
        stdout.print("Loop iteration: {}\n", .{i}) catch {};
    }

    var i: usize = 0;
    while (i < bar) : (i += 1) {
        if (i == 5) {
            break;
        }
        stdout.print("While loop iteration: {}\n", .{i}) catch {};
    }

    // Switch statement
    const value = 2;
    switch (value) {
        1 => stdout.print("Value is 1\n", .{}) catch {},
        2 => stdout.print("Value is 2\n", .{}) catch {},
        else => stdout.print("Value is something else\n", .{}) catch {},
    }

    // Functions
    const result = add(foobar, bar);
    stdout.print("Addition result: {}\n", .{result}) catch {};

    // Error handling
    var res = try divide(10, 2);
    stdout.print("Division result: {}\n", .{res}) catch {};
    res = divide(10, 0) catch {
        stdout.print("Caught divide by zero error\n", .{}) catch {};
        return;
    };
}

// Function definition
fn add(a: i32, b: i32) i32 {
    return a + b;
}

// Function with error handling
fn divide(a: i32, b: i32) !i32 {
    if (b == 0) {
        return error.DivideByZero;
    }
    return a / b;
}

// Custom error type
const error = struct {
    DivideByZero: void,
};

// Struct and enum
const CustomType = struct {
    field1: i32,
    field2: []const u8,
};

const Colors = enum {
    red,
    green,
    blue,
};

fn useStruct() void {
    var instance = CustomType{
        .field1 = 7,
        .field2 = "example",
    };
    var color = Colors.red;
}


